### PR TITLE
Adding notebook task type + small add ons

### PR DESCRIPTION
**Pull request by MaximHammer**
## Summary

- **Notebook task type** (`--task-type notebook`): Generates `notebook_task` entries instead of native `dbt_task`, enabling base environment support and environment variables on serverless compute. Uses a reusable runner notebook (`run_dbt_command.py`) that invokes dbt via the `dbtRunner` Python API with runtime token injection — same auth mechanism as native dbt_task.

- **Cold start improvements via base environments**: Documents and supports workspace base environments for serverless notebook tasks. Pre-caching `dbt-databricks` in a base environment skips package installation entirely, resulting in significantly faster cold starts.

- **Task naming improvements**: Models use plain names (e.g. `daily_orders` instead of `model_dbt_test_project_daily_orders`), tests use `<model>_tests` format (e.g. `daily_orders_tests`), seeds and snapshots keep their type prefix. Hash suffixes from generic dbt tests are stripped.

- **Dependency resolution improvements**: When tests are enabled, downstream models depend on upstream model **tests** rather than the model itself — ensuring data quality gates are enforced in the DAG. When `--run-tests false`, models depend directly on upstream models as before.

- **Update README** with new arguments, job template, notebook task docs, base environment usage

- **Update 2026-04-23**, addressed PR comments + using argparse.BooleanOptionalAction to address boolean bug existing prior

## Test plan
Ran full flow on test project with multiple model types & tests. Verified results in Databricks UI. Also ran basic tests locally.
[Full doc](https://docs.google.com/document/d/1jI1hEUc4eGHDDPjauWlGcEo0Moc32slcF2U4cr-snSo/edit?tab=t.0) with Links, Screenshots & Results

Also additionally test a custom run on jobs compute (SQL Commands still go through SQL Serverless)
image

- **Update 2026-04-23** - again tested full project parsing and run with example project consisting of various seeds, models, snapshots, tests

image
image

### CODE REVIEW

File: src/databricks_dbt_factory/Utils.py [hunk 1]
```diff
@@ -1,11 +1,10 @@
...
-    Args:
-        dbt_node_full_name (str): The name of the node.
+    resource_type, _, node_name = parts[0], parts[1], parts[2]
 
-    Returns:
-        str: The generated task key.
-    """
-    return dbt_node_full_name.replace('.', '_')
+    if resource_type == "model":
+        return node_name
```
Comment (mwojtyczka) [inline src/databricks_dbt_factory/Utils.py]: Stripping the package segment risks task_key collisions. Two different dbt packages with a model named `foo` (`model.pkg_a.foo` and `model.pkg_b.foo`) both resolve to `foo` — the second task either silently overwrites the first in a dict-based serializer or fails with a duplicate-key error in Databricks. At minimum, detect collisions and fail loudly; ideally, fall back to including the package when there's ambiguity.

File: src/databricks_dbt_factory/DbtFactory.py [hunk 2]
```diff
@@ -45,41 +28,55 @@ def create_tasks_and_update_job_spec(
...
-        Args:
-            dbt_manifest (dict): The DBT manifest content.
+        models_with_tests: set[str] = set()
+        if tests_enabled:
+            for node_info in dbt_nodes.values():
+                if node_info['resource_type'] != 'test':
+                    continue
+                for dep in node_info.get('depends_on', {}).get('nodes', []):
+                    if dep.startswith('model.') and dep in dbt_nodes:
+                        models_with_tests.add(dbt_nodes[dep]['name'])
```
Comment (mwojtyczka) [inline src/databricks_dbt_factory/DbtFactory.py]: `dbt_nodes[dep]['name']` drops the package name, mirroring the collision risk in `generate_task_key`. Two models from different packages with the same short name merge into a single `<name>_tests` entry and their tests are bundled together. Consider keying `models_with_tests` by the full node path (or the task_key from `generate_task_key`) so the aggregation is unambiguous.

File: src/databricks_dbt_factory/DbtFactory.py [hunk 3]
```diff
@@ -45,41 +28,55 @@ def create_tasks_and_update_job_spec(
...
 
-        Args:
-            dbt_manifest (dict): The DBT manifest content.
+        models_with_tests: set[str] = set()
+        if tests_enabled:
+            for node_info in dbt_nodes.values():
+                if node_info['resource_type'] != 'test':
+                    continue
+                for dep in node_info.get('depends_on', {}).get('nodes', []):
+                    if dep.startswith('model.') and dep in dbt_nodes:
```
Comment (mwojtyczka) [inline src/databricks_dbt_factory/DbtFactory.py]: Only tests whose `depends_on` includes a `model.*` node are aggregated. Tests defined on seeds, snapshots, or sources never produce a task and silently disappear from the generated job (previously each individual test node got its own task). I suppose we should broaden the filter to cover all testable resource types?

File: src/databricks_dbt_factory/DbtFactory.py [hunk 4]
```diff
@@ -45,41 +28,55 @@ def create_tasks_and_update_job_spec(
...
             task_key = generate_task_key(node_full_name)
             factory = self.task_factories[resource_type]
-
             task = factory.create_task(node_name, node_info, task_key)
+
+            if resource_type == 'model' and tests_enabled:
+                new_deps = []
+                for dep_key in (task.depends_on or []):
+                    if dep_key in models_with_tests:
+                        new_deps.append(f"{dep_key}_tests")
```
Comment (mwojtyczka) [inline src/databricks_dbt_factory/DbtFactory.py]: This relies on `dep_key` (a task_key from `generate_task_key`) happening to equal a plain node name in `models_with_tests`. It only works because `generate_task_key` for models returns the bare node name — change that convention and this breaks silently. Consider indexing `models_with_tests` by task_key so the contract between the two sets is explicit.

File: src/databricks_dbt_factory/DbtFactory.py [hunk 5]
```diff
@@ -45,41 +28,55 @@ def create_tasks_and_update_job_spec(
...
             tasks.append(task)
 
+        if tests_enabled:
+            test_factory = self.task_factories['test']
+            for model_name in sorted(models_with_tests):
+                task_key = f"{model_name}_tests"
+                dbt_deps = test_factory.get_dbt_deps_command(model_name)
+                commands = [dbt_deps] if dbt_deps else []
+                commands.append(
+                    f"dbt test --select {model_name}"
```
Comment (mwojtyczka) [inline src/databricks_dbt_factory/DbtFactory.py]: Two concerns here:

1. **Bypasses the factory pattern.** This reimplements `TestTaskFactory.create_task` inline — the constructed commands don't go through the factory, so future additions to the factory (e.g. extra selectors, tags, flags) won't apply to bundled test tasks. Consider adding a `create_bundled_task(model_name, node_full_name)` method on `TestTaskFactory` and calling that.

2. **Behavior change worth flagging.** Previously each dbt test node (`unique_x_id`, `not_null_x_id`, …) was its own Databricks task, giving per-test parallelism and a clear red task in the UI when a specific assertion fails. Now they're all collapsed into one `dbt test --select <model>` task. That's a reasonable simplification but it's a material UX change — worth documenting explicitly in the README and perhaps making the bundling configurable.

File: src/databricks_dbt_factory/DbtFactory.py [hunk 6]
```diff
@@ -45,41 +28,55 @@ def create_tasks_and_update_job_spec(
...
             task = factory.create_task(node_name, node_info, task_key)
+
+            if resource_type == 'model' and tests_enabled:
+                new_deps = []
+                for dep_key in (task.depends_on or []):
+                    if dep_key in models_with_tests:
+                        new_deps.append(f"{dep_key}_tests")
+                    else:
+                        new_deps.append(dep_key)
+                task = replace(task, depends_on=new_deps)
```
Comment (mwojtyczka) [inline src/databricks_dbt_factory/DbtFactory.py]: Minor: Forcing every model to depend on its upstream models' `_tests` task (rather than the models themselves) is a strong, non-opt-out opinion. Anyone who wants models to run without gating on tests — dev iteration, backfills, test triage — has to disable tests entirely with `--run-tests false`. Consider a separate flag (e.g. `--gate-on-tests`) so users can keep tests in the DAG without blocking downstream execution on them.

File: src/databricks_dbt_factory/main.py [hunk 7]
```diff
@@ -148,6 +173,13 @@ def parse_args():
         default=False,
     )
     args = parser.parse_args()
+
+    if args.task_type == "notebook" and not args.notebook_path:
+        parser.error("--notebook-path is required when --task-type is 'notebook'")
+
+    if args.job_cluster_key and args.environment_key != "Default":
+        parser.error("--job-cluster-key and --environment-key are mutually exclusive")
```
Comment (mwojtyczka) [inline src/databricks_dbt_factory/main.py:252]: `args.environment_key != "Default"` can't tell an unset default from an explicit `--environment-key Default`. A user passing `--job-cluster-key foo --environment-key Default` won't trip this error, and then `_base_spec` silently drops `environment_key` because `job_cluster_key` wins. Fix: use `default=None` on the argparse parameter and treat `None` as "not provided", falling back to `"Default"` at use-time in `DbtTaskOptions`.

File: src/databricks_dbt_factory/DbtTask.py [hunk 8]
```diff
@@ -56,31 +68,57 @@ class DbtTask:
 
     def to_dict(self) -> dict:
         """Converts the Task to a dictionary suitable for the job definition."""
+        if self.options.task_type == "notebook":
```
Comment (mwojtyczka) [inline src/databricks_dbt_factory/DbtTask.py]: String-typed `task_type` with magic values (`"dbt"` / `"notebook"`) means typos fail silently — `task_type="Notebook"` would just render as a dbt_task with no indication. Consider `typing.Literal["dbt", "notebook"]` on the dataclass field, or an `Enum` like the existing `DbtNodeTypes`.

File: src/databricks_dbt_factory/notebook/run_dbt_command.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    if profiles_directory:
+        args.extend(["--profiles-dir", profiles_directory])
+
+    print(f"Running: dbt {' '.join(args)}")
+    print("-" * 60)
+
+    result = runner.invoke(args)
+
+    if not result.success:
+        raise Exception(f"dbt command failed: dbt {' '.join(args)}\n{result.exception}")
```
Comment (mwojtyczka) [inline src/databricks_dbt_factory/notebook/run_dbt_command.py]: Bare `Exception` — prefer `RuntimeError` or a dedicated exception class so callers (and Databricks retries) can distinguish it from unrelated errors. Also, `result.exception` is only populated on some failure paths; for failed `dbt test`/`dbt run` runs the assertion/row errors are typically in `result.result`. Including that in the message makes the UI failure much easier to diagnose without drilling into logs.

File: src/databricks_dbt_factory/DbtTask.py [hunk 10]
```diff
@@ -56,31 +68,57 @@ class DbtTask:
...
-
+            dbt_task['project_directory'] = self.options.project_directory
         if self.options.schema:
-            spec['dbt_task']['schema'] = self.options.schema
-
-        if self.options.warehouse_id:  # not required if using "None (Manual) / Serverless"
-            spec['dbt_task']['warehouse_id'] = self.options.warehouse_id
+            dbt_task['schema'] = self.options.schema
+        if self.options.warehouse_id:
+            dbt_task['warehouse_id'] = self.options.warehouse_id
```
Comment (mwojtyczka) [inline src/databricks_dbt_factory/DbtTask.py:123]: When `task_type="notebook"`, `warehouse_id`, `schema`, and `catalog` on `DbtTaskOptions` are silently ignored (they only apply in `_to_dbt_dict`). If a user sets `--warehouse-id` together with `--task-type notebook`, they'll get no error but no warehouse binding either. Worth either a validation error in `main.py` or at least a log warning.

**Review by mwojtyczka — COMMENTED** [review]
Nice feature — notebook task type is a pragmatic workaround for the serverless dbt_task limitations, and the runner notebook is clean. A few concerns worth addressing before merge:

**Correctness / edge cases**
- Task-key collisions when two dbt packages share a model name (`Utils.py`, `DbtFactory.py`).
- Test aggregation ignores tests on seeds/snapshots/sources — they silently disappear from the job.
- The mutually-exclusive check between `--environment-key` and `--job-cluster-key` has a hole.

**Design / API**
- Test bundling (`dbt test --select <model>`) is a meaningful behavior change from per-test tasks — worth calling out in the docs, since it trades per-test parallelism and granular failure signal for a simpler DAG.
- Making downstream models depend on upstream `<model>_tests` should probably be opt-in rather than forced whenever tests are enabled.
- The inline test-task construction in `DbtFactory._create_tasks` reimplements `TestTaskFactory` logic and bypasses the factory abstraction.

**Minor**
- Several docstrings were removed from `DbtFactory` and `Utils` — that's a documentation regression even if the code is self-explanatory now.
- `CLAUDE.md` in `.gitignore` should live in a global gitignore, not the project.
- Bare `Exception` in the runner notebook; prefer a narrower type and include more of the `dbtRunner` result in the error.

Inline comments below on the specifics.

**Review by mwojtyczka — CHANGES_REQUESTED** [review]
Thanks @MaximHammer for the PR. This is a great extension.
I left some comments. Pls address.
Also can you please sign your commits. Create gpg key and repush the commeits.

File: README.md [hunk 11]
```diff
@@ -52,40 +53,162 @@ pip install databricks-dbt-factory
...
+          base_environment: "/Workspace/Shared/envs/my_base_env.yaml"
+```
+
+Note: `client` and `base_environment` are mutually exclusive — use one or the other.
+
+## Generating native dbt tasks
+
 ```shell
 databricks_dbt_factory  \
-  --dbt-manifest-path tests/test_data/manifest.json \
```
Comment (mwojtyczka) [inline README.md:58]: It would be great to add contributing guide which will include how to test, which would use this command

File: src/databricks_dbt_factory/DbtFactory.py [hunk 12]
```diff
@@ -1,20 +1,14 @@
+from dataclasses import replace
+
 from databricks_dbt_factory import TaskFactory
 from databricks_dbt_factory.SpecsHandler import SpecsHandler
 from databricks_dbt_factory.DbtTask import DbtTask
 from databricks_dbt_factory.Utils import generate_task_key
 
 
 class DbtFactory:
-    """A factory for generating Databricks job definitions from DBT manifests."""
```
Comment (mwojtyczka) [inline src/databricks_dbt_factory/DbtFactory.py:8]: any reason to remove all the docstring? they are quite useful, pls revert

**Review by mwojtyczka — APPROVED** [review]
LGTM
